File: lotteryticket/src/main/java/com/wuhai/lotteryticket/widget/popupwindow/PositionCanChangePopupWindow.java
package com.wuhai.lotteryticket.widget.popupwindow;

import android.content.Context;
import android.view.Gravity;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup.LayoutParams;
import android.widget.PopupWindow;

import com.wuhai.lotteryticket.R;
import com.wuhai.lotteryticket.ui.activity.LotteryCreateActivity;
import com.wuhai.lotteryticket.utils.LogProxy;

import java.lang.reflect.Field;

/**
*
 * 跟随点击或者长按位置  弹PopupWindow
* @author wuhai
* create at 2016/2/26 15:55
*/
public class PositionCanChangePopupWindow {

	private Context mContext;
	private View mParent;
	private View mContentView;
	private PopupWindow mWindow;
	private View mDelete,mEditTv,mSaveTv;

	private IFloatingOperation mOperCallback;

	public PositionCanChangePopupWindow(Context context, View parent) {
		mContext = context;
		mContentView = View.inflate(mContext, R.layout.popwindow_position, null);
		mWindow = new PopupWindow(mContentView, LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
		mWindow.setAnimationStyle(android.R.style.Animation_Toast);
		mParent = parent;
		mDelete = mContentView.findViewById(R.id.delete_tv);
		mEditTv = mContentView.findViewById(R.id.edit_tv);
		mSaveTv = mContentView.findViewById(R.id.save_tv);

		mDelete.setOnClickListener(mClickListener);
		mEditTv.setOnClickListener(mClickListener);
		mSaveTv.setOnClickListener(mClickListener);
	}

	private void setRelativeLayoutBackGround(boolean topOf) {
		if (topOf) {
			mContentView.setBackgroundResource(R.drawable.note_arrow_down_empty);
		} else {
			mContentView.setBackgroundResource(R.drawable.note_arrow_up_empty);
		}
	}

	/**
	 * ※ 值操作要把data或者postion 传过来 根据想法改动回传接口
	 *
	 * downX - 控件高度 < 状态栏高度  朝下；反之 朝上
	 * downX - 控件宽度/2 < 0  值为0;反之 正常差值
	 * @param downX
	 * @param downY
	 */
	public void showPopupWindow(float downX, float downY) {
		mContentView.measure(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
		int top = (int) (downY - mContentView.getMeasuredHeight());
		LogProxy.e(LotteryCreateActivity.TAG,"FootprintsPopupWindow::top:"+top+"="+downY+"-"+mContentView.getMeasuredHeight());
		boolean topOf = true;
		LogProxy.e(LotteryCreateActivity.TAG,"FootprintsPopupWindow::getStatusHeight:"+getStatusHeight(mContext));
		if (top < getStatusHeight(mContext)) {
			topOf = false;
			top = (int) downY;
		}
		setRelativeLayoutBackGround(topOf);
		int left = (int) (downX - mContentView.getMeasuredWidth() / 2);
//		System.out.println("FootprintsPopupWindow::left:"+left);
		mWindow.showAtLocation(mParent, Gravity.NO_GRAVITY, left < 0 ? 0 : left, top);
	}

	public void hide() {
		mWindow.dismiss();
	}

	/**
	 * Status bar是launcher主界面上面提示当前状态（电池，网络，蓝牙等等）的一个栏
	 * 获取状态栏高度
	 * @param context
	 * @return
	 */
	private int getStatusHeight(Context context){
		int height = 0;
		try {
			@SuppressWarnings("rawtypes")
            Class c = Class.forName("com.android.internal.R$dimen");
			Object obj = c.newInstance();
			Field field = c.getField("status_bar_height");
			int x = Integer.parseInt(field.get(obj).toString());
			height = context.getResources().getDimensionPixelSize(x);
		} catch (Exception e) {
			e.printStackTrace();
		}
		return height;
	}

	/**
	 * PopupWindow 是否正在显示
	 * @return
	 */
	public boolean isShowing(){
		if(mWindow == null){
			return false;
		}
		return mWindow.isShowing();
	}


	/**
	 * 回调接口
	 * @param l
	 */
	public void setFloatingOperation(IFloatingOperation l){
		mOperCallback = l;
	}

	final OnClickListener mClickListener = new OnClickListener() {
		@Override
		public void onClick(View v) {
			if(mOperCallback == null){
				return;
			}
			switch (v.getId()) {
				case R.id.edit_tv:
					mOperCallback.onEdit();
					break;
				case R.id.delete_tv:
					mOperCallback.onDelete();
					break;
				case R.id.save_tv:
					mOperCallback.onSave();
					break;
				default:
					break;
			}
			hide();
		}
	};

	public interface IFloatingOperation {
		void onEdit();
		void onDelete();
		void onSave();
	}
}

File: lotteryticketkotlin/src/main/java/com/wuhai/demo/lotteryticketkotlin/widget/popupwindow/PositionCanChangePopupWindow.kt
package com.wuhai.demo.lotteryticketkotlin.widget.popupwindow

import android.content.Context
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.PopupWindow
import com.wuhai.demo.lotteryticketkotlin.R
import com.wuhai.demo.lotteryticketkotlin.utils.LogProxy.e

/**
 *
 * 跟随点击或者长按位置  弹PopupWindow
 * @author wuhai
 * create at 2016/2/26 15:55
 */
class PositionCanChangePopupWindow(private val mContext: Context, parent: View) {
    private val mParent: View
    private val mContentView: View
    private val mWindow: PopupWindow?
    private val mDelete: View
    private val mEditTv: View
    private val mSaveTv: View
    private var mOperCallback: IFloatingOperation? = null
    private fun setRelativeLayoutBackGround(topOf: Boolean) {
        if (topOf) {
            mContentView.setBackgroundResource(R.drawable.note_arrow_down_empty)
        } else {
            mContentView.setBackgroundResource(R.drawable.note_arrow_up_empty)
        }
    }

    /**
     * ※ 值操作要把data或者postion 传过来 根据想法改动回传接口
     *
     * downX - 控件高度 < 状态栏高度  朝下；反之 朝上
     * downX - 控件宽度/2 < 0  值为0;反之 正常差值
     * @param downX
     * @param downY
     */
    fun showPopupWindow(downX: Float, downY: Float) {
        mContentView.measure(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        var top = (downY - mContentView.measuredHeight).toInt()
//        e(LotteryCreateActivity.TAG, "FootprintsPopupWindow::top:" + top + "=" + downY + "-" + mContentView.measuredHeight)
        var topOf = true
//        e(LotteryCreateActivity.TAG, "FootprintsPopupWindow::getStatusHeight:" + getStatusHeight(mContext))
        if (top < getStatusHeight(mContext)) {
            topOf = false
            top = downY.toInt()
        }
        setRelativeLayoutBackGround(topOf)
        val left = (downX - mContentView.measuredWidth / 2).toInt()
        //		System.out.println("FootprintsPopupWindow::left:"+left);
        mWindow!!.showAtLocation(mParent, Gravity.NO_GRAVITY, if (left < 0) 0 else left, top)
    }

    fun hide() {
        mWindow!!.dismiss()
    }

    /**
     * Status bar是launcher主界面上面提示当前状态（电池，网络，蓝牙等等）的一个栏
     * 获取状态栏高度
     * @param context
     * @return
     */
    private fun getStatusHeight(context: Context): Int {
        var height = 0
        try {
            val c = Class.forName("com.android.internal.R\$dimen")
            val obj = c.newInstance()
            val field = c.getField("status_bar_height")
            val x = field[obj].toString().toInt()
            height = context.resources.getDimensionPixelSize(x)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return height
    }

    /**
     * PopupWindow 是否正在显示
     * @return
     */
    val isShowing: Boolean
        get() = mWindow?.isShowing ?: false

    /**
     * 回调接口
     * @param l
     */
    fun setFloatingOperation(l: IFloatingOperation?) {
        mOperCallback = l
    }

    val mClickListener = View.OnClickListener { v ->
        if (mOperCallback == null) {
            return@OnClickListener
        }
        when (v.id) {
            R.id.edit_tv -> mOperCallback!!.onEdit()
            R.id.delete_tv -> mOperCallback!!.onDelete()
            R.id.save_tv -> mOperCallback!!.onSave()
            else -> {
            }
        }
        hide()
    }

    interface IFloatingOperation {
        fun onEdit()
        fun onDelete()
        fun onSave()
    }

    init {
        mContentView = View.inflate(mContext, R.layout.popwindow_position, null)
        mWindow = PopupWindow(mContentView, ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        mWindow.setAnimationStyle(android.R.style.Animation_Toast)
        mParent = parent
        mDelete = mContentView.findViewById(R.id.delete_tv)
        mEditTv = mContentView.findViewById(R.id.edit_tv)
        mSaveTv = mContentView.findViewById(R.id.save_tv)
        mDelete.setOnClickListener(mClickListener)
        mEditTv.setOnClickListener(mClickListener)
        mSaveTv.setOnClickListener(mClickListener)
    }
}